docs/design/fabric-v2+/fpc-registration.puml

```diff
@@ -7,266 +7,198 @@
+  Enclave1 -> Enclave1 : Attestation <- attest(SPID, Attested_Data)
+  note left #FFaa00
+    quote is defined with reportbody = Hash(<Attested_Data>)
   end note
 
-  return ECC_S_PK, ECC_E_PK, sealed state, quote
-
-  return ECC_S_PK, ECC_E_PK, sealed state, quote, SIG_Org_1
+  note over Enclave1 #FFaa00
+    Ideally the enclave itself is able to retrieve the verifiable evidence
+    (i.e., the verification report, using EPID) and return it.
+    This additionally requires some ocalls and helper code in untrusted environment.
+    For convenience, the enclave now returns the attestation, and the evidence
+    (i.e., the verification report) is retrieved from untrusted environment.
+  end note
+  Enclave1 -> IAS ++: Attestation
+  return Evidence <- Verification_Report
+end group
 
-  |||
+  Enclave1 -> Enclave1 : Sealed_Credentials <- seal(CC_Parameters, Private_Keys)
 
-  Manager1 -> storage1 ++ : store sealed state
-  note right
-    sealed state is stored under a
-    unique identifier consisting of
-    CC_NAME, CC_VER, CC_ID, ECC_S_PK, ECC_E_PK
+  note over Enclave1
+    The constant string "CREATE" in the Verb field is meant to be used only here
+    (in the create call). A different constant string is expected to be used whenever
+    the Credentials prototobuf is reused (e.g., to re-attest the chaincode). This
+    allows to distinguish different calls. 
   end note
-  return
+  Enclave1 -> Enclave1 : Credentials <- <Verb="CREATE", Attested_Data, Evidence>
+  return Credentials, Sealed_Credentials
 
-  create entity IAS order 22
-  Manager1 -> IAS ++: send quote
-  note right
-    use IAS credentials and url provided by config
-  end note
-  note over Manager1 #FFaa00
-    **MST**: From an attestation generalization perspective i think there should be just a single
-    call ""GetAttestationProof"" or alike rather than a separate place where you get quote
-    and another where you get the registration report which requires some re-organization of calls.
-    A few observations on this:
-    - we do not strictly have to call the enclave via ECC. ECCManager could also directly call
-      enclave.so. this would make this creation and key-management potentially easier?
-    - the quote generation is not driven from within the enclave as currently shown above but from
-      the outside!
-    I think from a high-level there should be following three high-level calls either in
-    ECC or ECCMAnager depending on above:
-    - EnclaveCreate (key-gen & imprint of context, i.e., current create/binding flows
-    - EnclaveProof: gets org endorsement-sig, creates report, then quote
-      and, if epid, gets IAS report. 
-    - EnclavePersist: returns sealed state (important for security that this is called only after
-      EnclaveProof [which we might want to enforce using some state-machine?] )
-  end note
-  return ATTESTATION
+  return Credentials, Sealed_Credentials
 
-  |||
+  E_Manager -> E_Manager : store Sealed_Credentials
```

Comment:
I am wondering where this is stored .... with recent discussion what the manager actually is, in particular, if the manger is part of a peer or part of a client.
1. If the manager is part of the peer, store will dump the sealed credentials to peers local storage
1. if the manager is part of a client, where is it stored?

Reply:
i have the same question further up (in fact, i don't think it will actually work to have it at the client with the current flows ...)

Reply:
@g2flyer  agreed, but if the manager is part of the peer, we have the problem that the manager cannot invoke "register transactions" (as I commented here also somewhere)

Reply:
Nothing prevents you from doing a transaction on the peer?

Reply:
ok - I agree that nothing prevents you from setup a client (some client credentials are needed) on the same node where a peer runs. Luckily the `peer chaincode` command also provides all the tool one need to run this from a peer host.

I would like to keep the logical separation of client, peer, and orderer here. 

- Client: invokes chaincode (endorsements), assembles transactions based on endorsements, submit transactions, (a special client called admin) performs special mgn operations such as install chaincode/approve/createchannel/etc...

- Peer: executes chaincodes invoked by clients and produces an endorsement, maintains a ledger of blocks of transactions, verifies transactions received from an ordering service; 

- Orderer: receives transactions, append to block, broadcast block, access control 

With the separation in mind I see the manager as part of the peer. If I think about an implementation of that, the admin calls the create operation and receives back a blob (attestation, cc_params?!), which is then used as input when the admin calls register as "normal" ercc invocation.

Let's quickly play this through ... (we extend the peer cli with fpc functions ... maybe using the wrapper) and let the admin/client use this.

the admin calls ...
```
$ peer fpc create -cc_name MyChaincode -ch_name MyChannel --peerAddresses 
172.23.45:7050
```

create will return an `attestation.blob` or similar 
```
$ cat attestation.blob
Attestation, Evidence, CC_Parms ....
$ peer chaincode invoke -cc_name ERCC -ch_name MyChannel --peerAddresses 
172.23.45:7050 172.23.55:7050 172.23.65:7050 -input attestation.blob
```

This completes creation of an enclave.  

This is a two step action, very similar how a new channel is created. There the admin first submits a transaction to the orderer and gets back a block, which is then used as input at a peer with the join command.

Alternatively, if we think about adding client functionality to the manager and host it together with the peer, we add an additional requirement in the network deployment that every peer must be provisioned with "some client" credentials that allows to invoke ercc. Setting up a fabric network with multiple peers and clients is already complicated enough I think, so this approach would make it even harder.

Reply:
> I would like to keep the logical separation of client, peer, and orderer here.
> 
> * Client: invokes chaincode (endorsements), assembles transactions based on endorsements, submit transactions, (a special client called admin) performs special mgn operations such as install chaincode/approve/createchannel/etc...

I think we should be careful, though, not to mix client in a sense of humans (e.g., admins) and client programs as complement to a server program. E.g., the external builder is a client to the peer as well but it is clearly not the same kind of client as an admin working from his laptop.

> * Peer: executes chaincodes invoked by clients and produces an endorsement, maintains a ledger of blocks of transactions, verifies transactions received from an ordering service;
> * Orderer: receives transactions, append to block, broadcast block, access control

> With the separation in mind I see the manager as part of the peer. If I think about an implementation of that, the admin calls the create operation and receives back a blob (attestation, cc_params?!), which is then used as input when the admin calls register as "normal" ercc invocation.
> 
> Let's quickly play this through ... (we extend the peer cli with fpc functions ... maybe using the wrapper) and let the admin/client use this.
> 
> the admin calls ...
> 
> ```
> $ peer fpc create -cc_name MyChaincode -ch_name MyChannel --peerAddresses 
> 172.23.45:7050
> ```

I think we should place that under `peer lifecycle chaincode ...` as it is really part of the lifecycle?

> create will return an `attestation.blob` or similar
> 
> ```
> $ cat attestation.blob
> Attestation, Evidence, CC_Parms ....
> ```
> 
> next the admin uses a normal chaincode invocation ...

Why do you even want to expose the blob to the admin (user), this could (imho should) all be part of above call?  I see your example of channel creation, but there it is as there multiple ways to create the channel definition (e.g., seems that the imagined "production-way" is not configtxgen but build it from  https://github.com/hyperledger/fabric-config; there was recently some related discussion on the fabric ML ....).  I don't think this would apply in our case?

> ```
> $ peer chaincode invoke -cc_name ERCC -ch_name MyChannel --peerAddresses 
> 172.23.45:7050 172.23.55:7050 172.23.65:7050 -input attestation.blob
> ```

If we decide that the human should do two commands instead of one as i suggest above, i would do this still as a (from the admin perspective) management command in the same namespace as above `create`.    This hides the current implementation (using invoke) and allows seemless tighter integration in the future.

> This completes creation of an enclave.
> 
> This is a two step action, very similar how a new channel is created. There the admin first submits a transaction to the orderer and gets back a block, which is then used as input at a peer with the join command.
> 
> Alternatively, if we think about adding client functionality to the manager and host it together with the peer, we add an additional requirement in the network deployment that every peer must be provisioned with "some client" credentials that allows to invoke ercc. Setting up a fabric network with multiple peers and clients is already complicated enough I think, so this approach would make it even harder.

Reply:
for now the sealing part is not implemented, and it could be easily implemented as/with the create, storing the sealed blob at the return point. This is not worth any additional time.